testsuite/tests/typing-jkind-bounds/row.ml

```diff
@@ -0,0 +1,278 @@
+module type S = sig
+  type 'a polyvar = private [< `A of 'a | `B of int ref ]
+  type 'a record : immutable_data with 'a polyvar = { inner : 'a }
+end
+module type S2 = S with type 'a polyvar = [ `A of 'a ]
+module F(M : S2) = struct
+  let cross (x : int M.record @ contended) = use_uncontended x
+end
+(* CR layouts v2.8: This should be accepted *)
+[%%expect{|
+module type S =
+  sig
+    type 'a polyvar = private [< `A of 'a | `B of int ref ]
+    type 'a record = { inner : 'a; }
+  end
+module type S2 =
+  sig type 'a polyvar = [ `A of 'a ] type 'a record = { inner : 'a; } end
+module F :
+  functor (M : S2) -> sig val cross : int M.record @ contended -> unit end
+|}]
+
+
+(* Harder cases: row variables *)
+
+(* CR layouts v2.8: These are both correct, but we could probably infer a more precise kind for both. *)
+type ('a, 'b) t : immediate = [< `X | `Y of 'a] as 'b
+[%%expect{|
+Line 1, characters 0-53:
+1 | type ('a, 'b) t : immediate = [< `X | `Y of 'a] as 'b
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "[< `X | `Y of 'a ]" is value
+         because it's a polymorphic variant type.
+       But the kind of type "[< `X | `Y of 'a ]" must be a subkind of immediate
+         because of the definition of t at line 1, characters 0-53.
+|}]
+type ('a, 'b) u : immediate = [> `X | `Y of 'a] as 'b
+[%%expect{|
+Line 1, characters 0-53:
+1 | type ('a, 'b) u : immediate = [> `X | `Y of 'a] as 'b
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "[> `X | `Y of 'a ]" is value
+         because it's a polymorphic variant type.
+       But the kind of type "[> `X | `Y of 'a ]" must be a subkind of immediate
+         because of the definition of u at line 1, characters 0-53.
+|}]
+
+(* less-than rows *)
+
+let f (x : [< `A of int | `B of string] @ contended) =
+  use_uncontended x
+(* CR layouts v2.8: This should probably be accepted *)
+[%%expect{|
+Line 2, characters 18-19:
+2 |   use_uncontended x
+                      ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+module M : sig
+  type 'a t : immutable_data with 'a = private [< `A of 'a | `B of ('a * 'a) | `C ]
+end = struct
+  type 'a t = [ `C ]
+end
+[%%expect{|
+Line 2, characters 2-83:
+2 |   type 'a t : immutable_data with 'a = private [< `A of 'a | `B of ('a * 'a) | `C ]
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "[< `A of 'a | `B of 'a * 'a | `C ]" is value
+         because it's a polymorphic variant type.
+       But the kind of type "[< `A of 'a | `B of 'a * 'a | `C ]" must be a subkind of
+         immutable_data with 'a
+         because of the definition of t at line 2, characters 2-83.
+|}]
+
+(* Tunivar-ified row variables *)
+
+type t1 = { f : ('a : value mod portable). ([> `Foo of int] as 'a) -> unit }
+(* CR layouts v2.8: This should be accepted *)
```

Comment:
No, I don't think so. Not until we have kind-restricted row extensions, which we have not planned out.

On the other hand, it would be much more plausible to accept this with a `<` instead of a `>`.

Reply:
Commented.